modules/Link.js

```diff
@@ -62,6 +77,63 @@ const Link = React.createClass({
     }
   },
 
+  getInitialState() {
+    return {
+      isActive: isLinkActive(this.context.router, this.props)
+    }
+  },
+
+  componentWillMount() {
+    const { router } = this.context
+    if (router) {
+      this.updateIsActive()
+    }
+  },
+
+  componentDidMount() {
+    const { router } = this.context
+    if (router) {
+      this._unlisten = router.listen(() => {
+        if (this._unlisten) {
+          this.updateIsActive()
+        }
+      })
+    }
+  },
+
+  componentWillReceiveProps(nextProps) {
+    const { router } = this.context
+    if (router) {
+      if (
+        nextProps.to !== this.props.to ||
+        nextProps.onlyActiveOnIndex !== this.props.onlyActiveOnIndex ||
+        nextProps.activeClassName !== this.props.activeClassName ||
+        nextProps.activeStyle !== this.props.activeStyle
+      ) {
+        this.updateIsActive(nextProps)
+      }
+    }
+  },
+
+  componentWillUnmount() {
+    if (this._unlisten) {
+      this._unlisten()
+      this._unlisten = null
+    }
+  },
+
+  updateIsActive(props = this.props) {
+    const { router } = this.context
+    const isActive = isLinkActive(router, props)
+
+    // The code is written this way to avoid wasted
+    // setState() calls that get expensive in large trees.
```

Comment:
I understand, but we only call it once initially (which we would call anyway), and when the *related* props have changed (which again, we’d have to recompute anyway).

`setState()` however *can* be slow on big trees. I’ve seen 3x difference on large trees in other project solved by avoiding `setState` even despite further `shouldComponentUpdate` bailout. React places `setState` in a queue and marks subtrees as dirty so this might be relevant.